### PR TITLE
add output field name rfc

### LINKED ISSUE #302: Create a `specification/` for `.md` documents with what we agreed upon
I would like to propose that we create a directory `specification/` where we drop `.md` files that result from the discussions we have over google docs, so that they are given the proper visibility in this repository.

We would continue to use google docs to iterate on them. Once ready, I propose to PR them to this repo.

Reasons:
1. Licensing: the work on them imo should remain ASF-licensed
2. visibility: we have a simple place where we can refer to
3. easier to contribute: it is easy propose changes to those documents
4. audit: changes to those documents are properly versioned, specially after we "agreed" on them

### PR DESCRIPTION
**Pull request by houqp**
# Which issue does this PR close?

Keeping the process light since the project is still in a rapid development phase.

Closes #302.
Relates to https://github.com/apache/arrow-datafusion/pull/55.

Once this is reviewed and accepted, I will send the update invariant doc in a follow up PR.

 # Rationale for this change

Unblock PR #55.

# What changes are included in this PR?

First RFC for field name semantic and RFC process.

### CODE REVIEW

File: docs/rfcs/output-field-name-semantic.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+###  Field name rules
+
+* All field names MUST not contain relation qualifier.
+  * Both `SELECT t1.id` and `SELECT id` SHOULD result in field name: `id`
+* Function names MUST be converted to lowercase.
+  * `SELECT AVG(c1)` SHOULD result in field name: `avg(c1)`
+* Literal string MUST not be wrapped with quotes or double quotes.
+  * `SELECT 'foo'` SHOULD result in field name: `foo`
+* Operator expressions MUST be wrapped with parentheses.
+  * `SELECT -2` SHOULD result in field name: `(- 2)`
```
Comment (alamb) [inline docs/rfcs/output-field-name-semantic.md:35]: Should there really be a space between `-` and `2`?

```suggestion
  * `SELECT -2` SHOULD result in field name: `(-2)`
```

**Review by alamb — APPROVED** [review]
Thanks @houqp  looks great to me

For anyone else following along here is a google doc link to drafts of this content.

https://docs.google.com/document/d/1uviWavwEGD3qxwMk2AGkOgp6ENrvKGiMWQhHNbqPwhg/edit

File: docs/rfcs/output-field-name-semantic.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
+# Datafusion output field name semantic
```
Comment (Dandandan) [inline docs/rfcs/output-field-name-semantic.md:1]: Might be helpful to number/add a date to the RFCs? E.g.

`2021-05-25-output-field-name-semantic.md` 
or
`00001-output-field-name-semantic.md `

**Comment by jorgecarleitao** [comment]
So, a question that I would like to bring before committing to RFCs. I see two ways of approaching this: 

1. the delta approach: each RFC may change previous RFCs, becoming the "latest" or a mixture of
2. the specification approach: each change is a PR, and the combined result is the new "specification"

Both have benefits and downsides.

My opinion is that we should not use the RFC approach, and instead work under the "specification" model.

The reason is that, in my opinion, RFCs are hard to follow because they were made in a specific moment in time and no longer updated. If updated, there is a new RFC that does that, which requires some form of consolidation (e.g. RFCs have the term "amends", "superseded by", "revoked", PEP also).

My opinion is that, with PRs, git history and git blame, there is no need to store the "deltas" (RFCs) in the repository itself, and we should instead offer the consolidated, up-to-date picture of the specification.

This was the rational of the original issue, at least.

File: docs/rfcs/output-field-name-semantic.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+* Mostly compatible with Spark’s behavior except literal string handling.
+* Mostly backward compatible with current Datafusion’s behavior other than
+function name cases and parenthesis around operator expressions.
+
+###  Field name rules
+
+* All field names MUST not contain relation qualifier.
+  * Both `SELECT t1.id` and `SELECT id` SHOULD result in field name: `id`
+* Function names MUST be converted to lowercase.
+  * `SELECT AVG(c1)` SHOULD result in field name: `avg(c1)`
```
Comment (Dandandan) [inline docs/rfcs/output-field-name-semantic.md:31]: One concern I had when implementing something very similar to this RFC here `https://github.com/apache/arrow-datafusion/pull/280` is that renaming the expressions / introducing an alias makes the column available under that name in the whole query. IMO that can be confusing if you do use this implicit convention. Any idea what other engines / databases are doing here?

Comment (Dandandan) [reply docs/rfcs/output-field-name-semantic.md:31]: SQLite: image

File: docs/rfcs/output-field-name-semantic.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+| 1 | 7   | foo_bar   |
+
+
+## Alternatives
+
+Postgres's behavior is too simple. It defaults to `?column?` as the column name
+in many cases, which makes output less readable than other designs.
+
+MySQL and SQLite preserve user query input as the field name. This adds extra
+implementation and runtime overhead with little gain for end uers.
```
Comment (Dandandan) [inline docs/rfcs/output-field-name-semantic.md:228]: What is the extra implementation / runtime overhead? I would say just printing the original (representation)  of expressions should be less work than getting rid of qualified names? Preserving casing is currently not something done by the parser - but qualified names are maintained.

File: docs/rfcs/output-field-name-semantic.md (hunk 3 above)
Comment (Dandandan) [reply docs/rfcs/output-field-name-semantic.md:31]: (So it looks like this is a common practice). Confusing - yes - but maybe not a hige concern, as users might not often rely on this as it is quite unconventional.

Comment (houqp) [reply docs/rfcs/output-field-name-semantic.md:31]: @Dandandan I documented a survey of behavior from mysql/sqlite/postgres/spark in this doc as well, for example: https://github.com/houqp/arrow-datafusion/blob/qp_rfc/docs/rfcs/output-field-name-semantic.md#function-with-operators. Basically mysql and sqlite use the raw user query as the column name, postgres throws in the towel and just use `?column?` for everything while spark SQL constructs the column name based on the expression.

I picked Spark's behavior because it's the one that is the closest to what we had at the time. But since you already implemented mysql and sqlite's behavior since then, i am happy to update the doc to account for that. In this case, we need two sets of rules, one for SQL queries, which is to just reuse what's provided in the query. The other one for dataframe queries, which is what I outlined in this doc.

File: docs/rfcs/output-field-name-semantic.md (hunk 1 above)
Comment (houqp) [reply docs/rfcs/output-field-name-semantic.md:35]: This came from the rule below: `Operator and operand MUST be separated by spaces.`. I don't have a strong opinion on this. Picked this because this is what spark does today. We can also change the rule below to require no spaces between operator and operands.

File: docs/rfcs/output-field-name-semantic.md (hunk 4 above)
Comment (houqp) [reply docs/rfcs/output-field-name-semantic.md:228]: When I wrote this, I was mostly thinking about attaching the extra alias expression node and having to traverse this node multiple times during plan optimization. But honestly, I don't think this overhead matters in practice. It's just one more thing we need to do compared to not preserving the raw query input.

Getting rid of the qualifier during physical planning is just one line change in the function that returns physical field name for a logical column.

**Comment by houqp** [comment]
@jorgecarleitao I agree with you, I also don't want this thing to get too formal and create unnecessary frictions whenever we need to update it. My original intent was a hybrid model of RFCs and Specs. Where we allow minor and backwards compatible changes to RFC directly (tracked through git), but require new RFCs for bigger changes. I will convert it to specs so people can walk in with the right expectation. We can come back to the RFC models in the future if we want to get serious about proposing changes.

**Comment by jorgecarleitao** [comment]
Thanks @houqp ! I am 100% with you here.

Wrt to the content itself, I am a big +1 here. We could discuss the finer print but my approval reads as: this is a fantastic first start; we may need some more iterations here; I would like to avoid committing to the exact wording at this point and instead allow further iterations on the existing text (i.e. without requiring another proposal / RFC, long text, and/or a way to consolidate texts.

**Comment by alamb** [comment]
Looks like we just need a RAT (apache copyright statement) to get a clean CI run and merge it. Perhaps we can add some section to the developer's guide once implemented with the "here is how output names are created" with a link to this RFC for the rationale

**Comment by alamb** [comment]
Looks like we just need a RAT (apache copyright statement) to get a clean CI run and merge it. Perhaps we can add some section to the developer's guide once implemented with the "here is how output names are created" with a link to this RFC for the rationale

**Comment by houqp** [comment]
@alamb @jorgecarleitao @Dandandan I reorganized everything to better align with the specification model. Could you take another look to see if there is anything you would like me to change?

**Review by jorgecarleitao — APPROVED** [review]
Beautiful. Thank you so much for your patience and work here. 💯

File: DEVELOPERS.md [hunk 5]
```diff
@@ -94,6 +94,19 @@ can be displayed. For example, the following command creates a
...
+We formalize Datafusion semantics and behaviors through specification
+documents. These specifications are useful to be used as references to help
+resolve ambiguities during development or code reviews.
+
+You are also welcome to propose changes to existing specifications or create
+new specifications as you see fit.
+
+Here is the list current active specifications:
+
+* [Output field name semantic](docs/specification/output-field-name-semantic.md)
```
Comment (alamb) [inline DEVELOPERS.md:108]: 👍

**Review by alamb — APPROVED** [review]
Thanks @houqp  -- looks good to me

**Comment by alamb** [comment]
windows test failure is unrelated
